### DISCUSSION TITLE
TcpConnect as command for rules

### DISCUSSION THREAD

**Discussion by juliogonzalez** [arendst/Tasmota#15518] [Q&A] [answered]
Right now, `PING` can be used to create a watchdog, as documented at https://tasmota.github.io/docs/Rules/#watchdog-for-wi-fi-router.

Now that there's [`TcpConnect`](https://github.com/arendst/Tasmota/pull/14874), and it would be great if it could be used inside rules as well.

This could be useful in situations such as:
- The device does not reply to `PING`, but exposes services on TCP ports
- The device crashed, still replies to `PING`, but you want a restart if a service crashed

**Comment by sfromis** [comment, marked as answer]
What stops you from using that command within rules? Basically **all** Tasmota commands available in the console and so on, are also available from a rule. 

Another question is if that command is at all relevant to what you want to do...

**Reply by juliogonzalez** [reply]
> What stops you from using that command within rules? Basically **all** Tasmota commands available in the console and so on, are also available from a rule.

Interesting, since the documentation does not tell what the result for `TcpConnect` is (while [it does](https://tasmota.github.io/docs/Commands/#wi-fi) for `PING`), I assumed it could not really be used. I will compile a firmware with the TCP bridge enabled and will check.

> Another question is if that command is at all relevant to what you want to do...

Should be. Basically the idea is checking if I can't connect to a TCP port, and if not, restart the device.

**Reply by juliogonzalez** [reply]
> Another question is if that command is at all relevant to what you want to do...

Or maybe you are right?

My understanding by reading the [doc](https://tasmota.github.io/docs/Commands/#wi-fi) is that this commands opens a TCP connection against a given TCP port and IP address.

But even with `USE_TCP_BRIDGE` enabled at build time, I can't get it to work:
```
19:36:34.513 CMD: ping4 192.168.1.1
19:36:34.532 MQT: stat/tasmota1/RESULT = {"Ping":"Done"}
19:36:38.771 MQT: tele/tasmota1/RESULT = {"Ping":{"192.168.1.1":{"Reachable":true,"IP":"192.168.1.1","Success":4,"Timeout":0,"MinTime":78,"MaxTime":79,"AvgTime":78}}}
```
```
19:36:43.815 CMD: tcpconnect 80,192.168.1.1
19:36:43.832 MQT: stat/tasmota1/RESULT = {"Command":"Error"}
```
Even if `192.168.1.1` has port 80 enabled:
```
$ telnet 192.168.1.1 80
Trying 192.168.1.1...
Connected to 192.168.1.1.
Escape character is '^]'.
```
Am I misunderstanding what `TcpConnect` does?

**Reply by sfromis** [reply]
As I understand it, the command `TCPConnect` is part of of the [Serial to TCP Bridge](https://tasmota.github.io/docs/Serial-to-TCP-Bridge/) feature, not something you can use in isolation. As explained in the thread you linked, the point is to also allow outbound connections, instead of expecting the bridge to wait for inbound connections. The bridge is only enabled with you also configure the relevant pins for the serial port you want to bridge.

The command [WebQuery](https://tasmota.github.io/docs/Commands/#webquery) or `WebSend` may be more relevant to you, since you refer to port 80.

**Reply by juliogonzalez** [reply]
> As I understand it, the command `TCPConnect` is part of of the [Serial to TCP Bridge](https://tasmota.github.io/docs/Serial-to-TCP-Bridge/) feature, not something you can use in isolation. As explained in the thread you linked, the point is to also allow outbound connections, instead of expecting the bridge to wait for inbound connections. The bridge is only enabled with you also configure the relevant pins for the serial port you want to bridge.

Not clear to me from the chunk of the doc (thread) I linked, specially because that's the `Wifi` section where `Ping` is also included. But seems you are right :-)

> The command [WebQuery](https://tasmota.github.io/docs/Commands/#webquery) or `WebSend` may be more relevant to you, since you refer to port 80.

Truth to be told, I used 80 as I wanted to provide a simple example using a FTTH router that doesn't provide SSH or Telnet, but ideally it should be possible to use any TCP port

Right now I have an Arduino Yun checking if a device is listening at port 22, and restarting it otherwise, but gladly get rid of it and replace it with some smartplug using tasmota.

That's what I thought `TCPConnect` would do (but clearly I was wrong :-))

**Reply by sfromis** [reply]
If you want to have a more general TCP interface with Tasmota on a smart switch, you get "more" if it is ESP32-based, where you also have the Berry programming language. Of course, I don't know if it would work out in your specific use case.

**Comment by pkkrusty** [comment]
tcpconnect worked for me after building with `#define USE_TCP_BRIDGE` , but first had to start the server with `tcpstart <port>`

```
20:30:22.048 CMD: tcpstart 1234
20:30:22.050 SRC: WebConsole from 192.168.1.103
20:30:22.051 CMD: Grp 0, Cmd 'TCPSTART', Idx 1, Len 4, Pld 1234, Data '1234'
20:30:22.053 TCP: Starting TCP server on port 1234
20:30:22.057 MQT: living-room-speakers/stat/RESULT = {"TCPStart":"Done"}
20:30:54.612 CMD: tcpconnect 3000,192.168.1.14
20:30:54.615 CMD: Grp 0, Cmd 'TCPCONNECT', Idx 1, Len 17, Pld 3000, Data '3000,192.168.1.14'
20:30:54.617 TCP: Connecting to 192.168.1.14 on port 3000
20:30:54.625 TCP: connected!
```

**Reply by bepluy** [reply]
But how can you send anything after get connected ? 
I have some device on my network which is waiting simple value (temperature) on tcp port.
Can you give me any example of sending anything after getting connected to tcp server ?

**Reply by sfromis** [reply]
The command `TCPConnect` is not part of some sort of general TCP feature, but specifically tied to bridging serial/modbus traffic to TCP. Therefore no command for sending TCP traffic is available. For such wishes, use Tasmota on ESP32 where you have the Berry language available, including support for coding around TCP connections.

If you want to experiment, you could try if you can send serial data on one pin, and get that connected to the serial bridge. Not sure if this would be a waste of time or not :wink:

**Reply by bepluy** [reply]
ESP32 is a bit expensive for some cheap temp/hum room sensors. The main idea is to get temp value from internal ds18b20 or sht30 and send it over wifi to some heating device, which understands simple temp value on tcp port.
So it would be very useful to have some raw "TcpSend" command as "WebSend" we have now.

**Reply by barbudor** [reply]
Esp32c3 is competitive with esp8266
Adapting protocols is also easy to do in backend such as node red
